### PR TITLE
[WIP] Feature/integration

### PR DESCRIPTION
**Pull request by MizukiSonoko**
## What is this pull request?
This is integration of all

## Why do you implement it? Why do we need this pull request?
We must release Iroha!

## Details/Features
List of features / major commits
- [ ] ametsuchi
- [ ] simulator 
- [ ] executor/validator
- [ ] consensus
- [x] ordering service
- [ ] peer service
- [ ] logger
- [ ] proto -> model

## Functional 
- [x] Receive transaction
- [ ] Validate / Execute primitive logic
- [x] Consensus building
- [ ] Store transaction with block chain
- [ ] Store entity 
 
## Additional dependencies (optional)
All

### CODE REVIEW

File: docker/docker-compose-ametsuchi-dev.yml [hunk 1]
```diff
@@ -64,7 +64,7 @@ services:
         ipv4_address: 192.168.9.4
```
Comment (takemiyamakoto) [inline docker/docker-compose-ametsuchi-dev.yml]: Putting a real IP here may not be great ^^

File: irohad/api/command_service.cpp [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+    void receive(
+        std::function<void(const iroha::protocol::Transaction&)> const func) {
+      dispatchToOrdering = func;
+    }
+
+    grpc::Status CommandService::Torii(grpc::ServerContext* context,
+                                       const ::iroha::protocol::Transaction* request,
+                                       ::iroha::protocol::ToriiResponse* response){
+      // TODO: Use this to get client's ip and port.
```
Comment (takemiyamakoto) [inline irohad/api/command_service.cpp]: This seems important^^

File: irohad/consensus/consensus_client.cpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+    if (status.ok()) {
+      return ack;
+    } else {
+      console->error("SendVote RPC failed: {}", status.error_details());
+      throw std::system_error();  // TODO: we need good exception design
+    }
+  }
+
+  Ack ConsensusClient::SendCommit(Commit *commit) {
```
Comment (takemiyamakoto) [inline irohad/consensus/consensus_client.cpp]: Maybe const Commit *commit is better?

File: irohad/consensus/consensus_service.cpp [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "consensus_service.hpp"
+
+namespace iroha {
+
+  Status ConsensusService::SendProposal(ServerContext *context,
```
Comment (takemiyamakoto) [inline irohad/consensus/consensus_service.cpp]: We should use const where we can.

File: irohad/consensus/state/leader.cpp [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+#include "leader.hpp"
+#include <future>
+
+static auto console = spdlog::stdout_color_mt("leader");
+
+namespace iroha {
+  void Leader::on_proposal(Proposal *proposal) {
+    console->info("Proposal handled");
+
+    switch (state_) {
```
Comment (takemiyamakoto) [inline irohad/consensus/state/leader.cpp]: Regular if-then is probably less error prone than a switch statement.

File: irohad/consensus/state/leader.cpp [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    }
+  }
+
+  Role Leader::self() { return Role::LEADER; }
+
+  void Leader::on_commit(Commit *commit) {
+    console->info("Commit handled");
+
+    {
+      // check if we received commit for last proposal
```
Comment (takemiyamakoto) [inline irohad/consensus/state/leader.cpp]: If something is not completed, let's add a TODO

File: irohad/consensus/state/member.cpp [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+  void Member::on_vote(Vote *vote) { console->info("Vote handled"); }
+
+  void Member::on_abort(Abort *abort) { console->info("Abort handled"); }
+
+  void Member::on_proposal(Proposal *proposal) {
+    console->info("Proposal handled");
+  }
+
+  std::string Member::ledgerToString() {
```
Comment (takemiyamakoto) [inline irohad/consensus/state/member.cpp]: Can we perhaps make this less ugly? Creating a stringutils library of functions might be a good choice.

File: irohad/consensus/state/proxy_tail.cpp [hunk 8]
```diff
@@ -15,21 +15,15 @@
...
-      rxcpp::observable<model::Proposal> on_proposal() override;
-     private:
-      rxcpp::subjects::subject<model::Proposal> proposals_;
-    };
-  }//namespace ordering
-}// namespace iroha
 
-#endif //IROHA_ORDERING_SERVICE_STUB_HPP
+  void ProxyTail::on_vote(Vote *vote) {
+    console->info("Vote handled");
```
Comment (takemiyamakoto) [inline irohad/consensus/state/proxy_tail.cpp]: If there is something not finished, please add a TODO.

File: irohad/consensus/state/validator.cpp [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+ */
+
+#include "validator.hpp"
+
+static auto console = spdlog::stdout_color_st("validator");
+
+namespace iroha {
+
+  void Validator::on_proposal(Proposal *proposal) {
+    // TODO
```
Comment (takemiyamakoto) [inline irohad/consensus/state/validator.cpp]: TODO: what?

File: irohad/consensus/state/validator.cpp [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+ * limitations under the License.
+ */
+
+#include "validator.hpp"
+
+static auto console = spdlog::stdout_color_st("validator");
+
+namespace iroha {
+
+  void Validator::on_proposal(Proposal *proposal) {
```
Comment (takemiyamakoto) [inline irohad/consensus/state/validator.cpp]: Can we use *const* keyword?

File: irohad/consensus/state/validator.cpp [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+    // TODO
+
+    // I don't like if-else/switch statements for states, can we do something
```
Comment (takemiyamakoto) [inline irohad/consensus/state/validator.cpp]: I generally prefer if-then-else to switch statements.

File: irohad/consensus/sumeragi.cpp [hunk 12]
```diff
@@ -143,72 +113,17 @@ namespace consensus {
      */
 
     int getNextOrder() {
-      thread_local int
-          currentProxyTail = static_cast<int>(getNumValidatingPeers()) - 1;
-      if (currentProxyTail >= peer_service::monitor::getActivePeerSize()) {
-        return -1;
-      }
-      return currentProxyTail++;
+      return 0;
```
Comment (takemiyamakoto) [inline irohad/consensus/sumeragi.cpp]: What does this mean? We also use 0th as leader?

If something is not finished add a TODO

File: irohad/main/irohad.cpp [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+    // end
+
+    console_->info("Current order: {}", state_->peerService->orderToString());
+    console_->debug("started");
+
+    loop_->run();
+  }
+
+  void Irohad::bind_callbacks() {
+    // temp solution
```
Comment (takemiyamakoto) [inline irohad/main/irohad.cpp]: TODO: keyword can help for grep

File: irohad/main/irohad.cpp [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+                                              auto& s) {
+      c->error("{}", e.what());
+
+      // TODO: we need better exception management
+      throw std::system_error();
+    });
+
+    std::shared_ptr<Member> state{state_};
+
+    server_.on<Proposal*>(
```
Comment (takemiyamakoto) [inline irohad/main/irohad.cpp]: The following code here is not only hard to read, but there are no comments explaining what it does.

File: irohad/main/irohad.cpp [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+      // simulate client
+      // creates separate thread which pushes transaction in queue
+      std::thread([os{this->orderingService}]() {
+        while (1) {
+          std::this_thread::sleep_for(std::chrono::milliseconds(1500));
+          os->simulate_one();
+        }
+      }).detach();
+
+    } else if (pos == 2 * ps->f() + 1) {
```
Comment (takemiyamakoto) [inline irohad/main/irohad.cpp]: Let's avoid logic like this as it is very fragile.

File: irohad/peer_service/PeerService.cpp [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "PeerService.hpp"
+
+namespace iroha {
+
+  NetworkNode::NetworkNode(std::string ip_, uint16_t port_,
```
Comment (takemiyamakoto) [inline irohad/peer_service/PeerService.cpp]: We should use *const* where we can.

File: irohad/peer_service/PeerService.cpp [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+  size_t PeerService::f() const { return (peers.size() - 1) / 3; }
+
+  size_t PeerService::position(ed25519::pubkey_t pub) {
+    size_t p = 0;
+    for (auto &&peer : peers) {
+      if (peer->pubkey == pub) {
+        return p;
+      }
+
+      p++;
```
Comment (takemiyamakoto) [inline irohad/peer_service/PeerService.cpp]: ++p

File: irohad/peer_service/PeerService.hpp [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+    size_t position(ed25519::pubkey_t pub);
+
+    /**
+     * Dummy function for "PeerService" run
+     */
+    void run();
+    void makeOrder();
+
+    // debug function
+    std::string orderToString() {
```
Comment (takemiyamakoto) [inline irohad/peer_service/PeerService.hpp]: We should add a stringUtils library to handle stuff like this

File: irohad/peer_service/self_state.cpp [hunk 19]
```diff
@@ -22,13 +22,15 @@ limitations under the License.
 #include <unistd.h>
 
 #include <crypto/crypto.hpp>
-
+#include <logger/logger.hpp>
 #include <iostream>
 #include <cstring>
 
 namespace peer_service {
   namespace self_state {
```
Comment (takemiyamakoto) [inline irohad/peer_service/self_state.cpp]: This file really needs more comments

File: irohad/torii/command_service.cpp [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+      *response = ToriiResponse();
+      response->set_code(ResponseCode::OK);
+      response->set_message("successfully dispatching to ordering.");
+    } else {
+      // TODO: Return validation failed message
+      *response = ToriiResponse();
+      response->set_code(ResponseCode::FAIL);
+      response->set_message("failed stateless validation.");
+    }*/
+    return grpc::Status::OK;
```
Comment (takemiyamakoto) [inline irohad/torii/command_service.cpp]: This is not very friendly to people trying to test.

File: irohad/torii/processor/transaction_processor.hpp [hunk 21]
```diff
@@ -21,24 +21,28 @@
...
 #include <model/transaction.hpp>
 #include <model/client.hpp>
+#include <validation/stateless/validator.hpp>
 
 namespace iroha {
   namespace torii {
 
     /**
-     * Transaction processor is interface with start point
+     * Transaction handler is interface with start point
```
Comment (takemiyamakoto) [inline irohad/torii/processor/transaction_processor.hpp]: Changing names like this should really be approved by me, as this is a core component of the system.

Generally I think TransactionHandler is an okay name, but everyone needs to know why the name changed.

File: test/scenario/test_network.cpp [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+
+#include <crypto/crypto.hpp>
+#include <main/irohad.hpp>
+#include <memory>
+#include <uvw/loop.hpp>
+
+using namespace iroha;
+using namespace ed25519;
+
+#define NPEERS 4
```
Comment (takemiyamakoto) [inline test/scenario/test_network.cpp]: we should try with varying numbers of peers

File: test/scenario/test_network.cpp [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  std::vector<std::shared_ptr<Irohad>> peers;
+  std::vector<std::thread> threads;
+};
+
+int main(){  // client thread
+  Network network;
+
+  while (1) {
```
Comment (takemiyamakoto) [inline test/scenario/test_network.cpp]: How does this ever end?

File: irohad/main/application.cpp [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+
+Irohad::Irohad() :
+    context(new Context()) {}
+
+void Irohad::run() {
+//  iroha::Irohad irohad;
+//  iroha::ametsuchi::StorageImpl ametsuchi;
+
+  // TODO replace with actual public private keys
+  auto seed = iroha::create_seed("some passphrase");
```
Comment (takemiyamakoto) [inline irohad/main/application.cpp]: Do we need to add real private keys now?

File: irohad/consensus/state/member.cpp (hunk 7 above)
Comment (Warchant) [inline irohad/consensus/state/member.cpp]: it was just for debug

File: docker/docker-compose-dev.yml [hunk 25]
```diff
@@ -42,7 +42,7 @@ services:
         ipv4_address: 192.168.9.3
```
Comment (takemiyamakoto) [inline docker/docker-compose-dev.yml]: Do we have to hard code this IP here?

File: libs/logger/logger.cpp [hunk 26]
```diff
@@ -15,4 +15,30 @@ limitations under the License.
...
 // This file is used to link dependencies library
-#include "logger.hpp"
\ No newline at end of file
+#include "logger.hpp"
+
+namespace logger {
+
+    // To hide spdlog
+    void setGlobalLogLevel(const level &l) {
+        switch (l) {
```
Comment (takemiyamakoto) [inline libs/logger/logger.cpp]: I don't like switch statements

File: irohad/peer_service/PeerService.cpp (hunk 17 above)
Comment (motxx) [inline irohad/peer_service/PeerService.cpp]: In c++, primitive types don't change any bytecodes. Using iterator or something may affect speed.

File: test/scenario/test_network.cpp (hunk 23 above)
Comment (l4l) [inline test/scenario/test_network.cpp]: If we doesn't want to end emulation that is fine. That should be replaced with `join` calls of related network threads

File: irohad/api/command_service.cpp [hunk 27]
```diff
@@ -0,0 +1,56 @@
...
+      (void)context;
```
Comment (motxx) [inline irohad/api/command_service.cpp]: It means avoiding warning that value is not used.
But for now, this file will completely be removed. Thanks.

**Comment by l4l** [comment]
That's fine with building but there's some undefined behavior in __torii_async_test__

File: irohad/ordering/OrderingService.hpp [hunk 28]
```diff
@@ -0,0 +1,54 @@
...
+
+#ifndef IROHA_ORDERINGSERVICE_HPP
+#define IROHA_ORDERINGSERVICE_HPP
+
+#include <spdlog/spdlog.h>
+#include <consensus/messages.hpp>
+#include <queue>
+#include <uvw.hpp>
+
+namespace ordering {
```
Comment (MizukiSonoko) [inline irohad/ordering/OrderingService.hpp]: In my opinion,
Generally namespace is same as directory/file name in irohad
`consensus/sumeragi.hpp` => `namespace consensus::sumeragi`;
if filename is same as directoryname, we abbreviates.
`foo/foo.hpp` => `namespace foo`;

In libs, first namespace is `iroha`
`libs/logger/logger.hpp` => `namespace iroha::logger`;

what do you think about?

Comment (takemiyamakoto) [inline irohad/ordering/OrderingService.hpp]: I agree with Mizuki. We only need to add iroha in libs I think.

Comment (l4l) [inline irohad/ordering/OrderingService.hpp]: I like that. Do you mind to note this somewhere (atlassian/wiki/contributing.md/somewhere else)?

**Comment by Warchant** [comment]
Consensus is not yet done

**Comment by neewy** [comment]
Fix commit name please

**Comment by lebdron** [comment]
`feature/consensus` was removed since it is outdated.
transaction processor map is already in develop.
logger will be added later.

**Pull request merged**